### ISSUE TITLE
Owin broke Crowbar

### ISSUE THREAD

**Issue by enamrik** [gravplats/crowbar#4]
As mentioned in [issue #3](https://github.com/mrydengren/crowbar/issues/3), Owin breaks Crowbar. The stacktrace from that issue is pretty clear that Owin is the cause. I spent sometime trying to figure out why and hoped to find a fix. So far things look grim. The problem arises in the following code from System.Web's HttpResponse:

```
public NameValueCollection Headers
{
  get
  {
    if (!(this._wr is IIS7WorkerRequest))
      throw new PlatformNotSupportedException(SR.GetString("Requires_Iis_Integrated_Mode"));
    if (this._headers == null)
      this._headers = new HttpHeaderCollection(this._wr, this, 16);
    return (NameValueCollection) this._headers;
  }
}
```

In order for Owin to create its context, it needs to grab the response headers. The problem is HttpResponse won't allow access to the headers if the current HttpWorkerRequest isn't of type IIS7WorkerRequest. Crowbar is calling HttpRuntime.ProcessRequest programmatically and passing in it's own custom HttpWorkerRequest, CrowbarHttpWorker. But CrobarHttpWorker inherits from SimpleWorkerRequest, not IIS7WorkerRequest and so the exception in the code snippet above gets thrown. It would have been an easy enough fix to simply have CrowbarHttpWorker inherit from IIS7WorkerRequest but IIS7WorkerRequest is sealed and internal. I've looked at other projects similar to Crowbar like CasiniDev but they use the same idea of calling HttpRuntime.ProcessRequest with a custom HttpWorkerRequest so we can assume they won't work with Owin either.

Owin is not required when using MVC5 but Microsoft is promoting Owin. They rewrote FormsAuthentication as an Owin Middleware and all of their security modules are written as Owin Middleware. I understand you're busy so I'd love to help out on getting Crowbar working with Owin. I'm fresh out of ideas but if you have any then please let me know and I'll take care of researching and spiking them out. Oh, and great tool by the way :-)

**Comment by mrydengren** [comment]
Hello, and thank you for your interest in Crowbar.

Unfortunately, I've no experience with OWIN.

I've previously experienced a similiar exception with response headers in Crowbar. But that issue got fixed in the following commit https://github.com/mrydengren/crowbar/commit/0077e58703dd487752e661bf3f431eda09d322e5.

The heart and soul of Crowbar is right here: https://github.com/mrydengren/crowbar/blob/master/src/Crowbar/Client.cs#L98. It might be that this approach won't work with OWIN. Is the custom HTTP worker (CrowbarHttpWorker) being called at all or are we missing some additional overrides.

I wish I could be more helpful. However, if time permitts I might spend a few hours on this issue during the holidays, but I won't make any promises.

**Comment by enamrik** [comment]
Thanks for responding. I looked at the commit you mentioned. The comment " Cannot read headers from response as it is not supported" lets me know that you ran into the exact same problem. Your fix was to implement a few overrides of HttpWorkerRequest and have ASP.NET pass you the response headers instead of reaching into HttpResponse.Headers and getting them yourself which would throw an exception (the same exception in the snippet from my original post). 

Owin on the other hand IS reading from the HttpResponse.Headers property. When running ASP.NET with IIS7+, no exception is thrown since HttpWorkerRequest is indeed an IIS7WorkerRequest. But when programmatically running ASP.NET through Crowbar, HttpWorkerRequest will not be an IIS7WorkerRequest and thus the exception is thrown. The maintainers of the Katana project (Microsoft's Owin implementation, which is what I mean when I say Owin) have no plans of supporting IIS6 so we can forget about them changing how Owin works. Owin is implemented as an HttpModule. I spent sometime today browsing HttpRuntime, HttpApplication and HttpContext looking for an interception point but so far I haven't found any.

It looks like I'm going to have to pass on Crowbar for my current project so no rush. But I'm on board to help in anyway I can and will continue looking for a fix in my spare time.

P.S. Once ASP.NET MVC is no longer coupled with System.Web and is entirely an Owin application, it will be pretty easy to write an integration testing framework for MVC without bothering with HttpRuntime and AppDomains. Here's to that day getting here sooner than later!
